pkg/kubectl/rollback.go

```diff
@@ -119,113 +115,132 @@ func (r *DeploymentRollbacker) Rollback(obj runtime.Object, updatedAnnotations m
 		return "", fmt.Errorf("failed to retrieve Deployment %s: %v", name, err)
 	}
 
+	templateForRevision, latestTemplate, err := deploymentRevision(deployment, r.c, toRevision)
+	if err != nil {
+		return "", err
+	}
 	if dryRun {
-		return simpleDryRun(deployment, r.c, toRevision)
+		return printTemplate(templateForRevision)
 	}
 	if deployment.Spec.Paused {
 		return "", fmt.Errorf("you cannot rollback a paused deployment; resume it first with 'kubectl rollout resume deployment/%s' and try again", name)
 	}
-	deploymentRollback := &extensionsv1beta1.DeploymentRollback{
-		Name:               name,
-		UpdatedAnnotations: updatedAnnotations,
-		RollbackTo: extensionsv1beta1.RollbackConfig{
-			Revision: toRevision,
-		},
+
+	// Skip if the revision already matches current Deployment
+	if equalIgnoreHash(templateForRevision, latestTemplate) {
+		return fmt.Sprintf("%s (current template already matches revision %d)", rollbackSkipped, toRevision), nil
 	}
-	result := ""
 
-	// Get current events
-	events, err := r.c.CoreV1().Events(namespace).List(metav1.ListOptions{})
-	if err != nil {
-		return result, err
+	// remove hash label before patching back into the deployment
+	delete(templateForRevision.Labels, appsv1.DefaultDeploymentUniqueLabelKey)
+
+	// compute deployment annotations
+	annotations := map[string]string{}
+	for k := range annotationsToSkip {
+		if v, ok := deployment.Annotations[k]; ok {
+			annotations[k] = v
+		}
 	}
-	// Do the rollback
-	// TODO: This is DEPRECATED. It should be updated. DaemonSets and StatefulSets implement rollback by
-	// patching using history (ControllerRevision data). Deployments should probably also implement
-	// rollback using a patch.
-	if err := r.c.ExtensionsV1beta1().Deployments(namespace).Rollback(deploymentRollback); err != nil {
-		return result, err
+	for k, v := range templateForRevision.Annotations {
```

Comment:
RS annotations should be copied (to the Deployment), not RS template annotations.